lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2569,18 +2569,78 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
     nominal->diagnose(diag::property_delegate_reserved_name);
   }
 }
+
 // SWIFT_ENABLE_TENSORFLOW
 /// Returns true if the given type conforms to `Differentiable` in the given
 /// module.
-static bool conformsToDifferentiableInModule(Type type, ModuleDecl *module) {
-  auto &ctx = module->getASTContext();
+static bool conformsToDifferentiable(TypeChecker &TC, Type type,
+                                     DeclContext *DC) {
   auto *differentiableProto =
-      ctx.getProtocol(KnownProtocolKind::Differentiable);
-  return LookUpConformanceInModule(module)(
-      differentiableProto->getDeclaredInterfaceType()->getCanonicalType(),
-      type, differentiableProto).hasValue();
+      TC.Context.getProtocol(KnownProtocolKind::Differentiable);
+  return TC.conformsToProtocol(
+      type, differentiableProto, DC,
+      ConformanceCheckFlags::Used | ConformanceCheckFlags::InExpression)
+          .hasValue();
 };
 
+// SWIFT_ENABLE_TENSORFLOW
+/// Creates a `AutoDiffParameterIndices` for the given function type,
+/// inferring all differentiation parameters.
+/// The differentiation parameters are inferred to be:
+/// - All parameters of the function type that conform to `Differentiable`.
+/// - If the function type's result is a function type, then also all
+///   parameters of the function result type that conform to `Differentiable`.
+static AutoDiffParameterIndices *
+inferDifferentiableParameters(TypeChecker &TC,
+                              AbstractFunctionDecl *AFD,
+                              GenericEnvironment *derivativeGenEnv) {
+  auto *functionType = AFD->getInterfaceType()->eraseDynamicSelfType()
+      ->castTo<AnyFunctionType>();
+  AutoDiffParameterIndicesBuilder builder(functionType);
+  SmallVector<Type, 4> allParamTypes;
+
+  // Returns true if the i-th parameter type is differentiable.
+  auto isDifferentiableParam = [&](unsigned i) -> bool {
+    if (i >= allParamTypes.size())
+      return false;
+    auto paramType = allParamTypes[i];
+    if (!paramType->hasTypeParameter())
+      paramType = paramType->mapTypeOutOfContext();
```

Comment:
These two lines are a no-op. mapTypeOutOfContext() substitutes type parameters, but you're only calling it if the type has no type parameters. You should just be able to delete this.

Reply:
Thanks! Done in https://github.com/apple/swift/pull/25324/commits/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13. I'll try to avoid this no-op pattern in the future.